nuxeo-mongodb-directory/src/main/java/org/nuxeo/directory/MongoDBSession.java

```diff
@@ -0,0 +1,349 @@
+
+    protected final Map<String, Field> schemaFieldMap;
+
+    public MongoDBSession(MongoDBDirectory directory) {
+        super(directory);
+        try {
+
+            client = MongoDBConnectionHelper.newMongoClient(directory.getDescriptor().getServerUrl());
+            dbName = directory.getDescriptor().getDatabaseName();
+            directoryName = directory.getName();
+            schemaName = directory.getSchema();
+            substringMatchType = directory.getDescriptor().getSubstringMatchType();
+            schemaFieldMap = directory.getSchemaFieldMap();
+
+        } catch (UnknownHostException e) {
+            throw new NuxeoException(e);
+        }
+    }
+
+    @Override
+    public MongoDBDirectory getDirectory() {
+        return (MongoDBDirectory) directory;
+    }
+
+    @Override
+    public DocumentModel getEntry(String id) throws DirectoryException {
+        return getEntry(id, true);
+    }
+
+    @Override
+    public DocumentModel getEntry(String id, boolean fetchReferences) throws DirectoryException {
+        if (!hasPermission(SecurityConstants.READ)) {
+            return null;
+        }
+        return directory.getCache().getEntry(id, this, fetchReferences);
+    }
+
+    @Override
+    public DocumentModelList getEntries() throws DirectoryException {
+        if (!hasPermission(SecurityConstants.READ)) {
+            return new DocumentModelListImpl();
+        }
+        return query(Collections.emptyMap());
+    }
+
+    @Override
+    public DocumentModel createEntry(Map<String, Object> fieldMap) throws DirectoryException {
+        checkPermission(SecurityConstants.WRITE);
+        String id = String.valueOf(fieldMap.get(getIdField()));
+        if (hasEntry(id)) {
+            throw new DirectoryException(String.format("Entry with id %s already exists", id));
+        }
+        try {
+            Document bson = MongoDBSerializationHelper.fieldMapToBson(fieldMap);
+            getCollection().insertOne(bson);
+            DocumentModel docModel = BaseSession.createEntryModel(null, schemaName, id, fieldMap, isReadOnly());
+            return docModel;
+        } catch (MongoWriteException e) {
+            throw new DirectoryException(e);
+        }
+    }
+
+    @Override
+    public void updateEntry(DocumentModel docModel) throws DirectoryException {
+
+        checkPermission(SecurityConstants.WRITE);
+        Map<String, Object> fieldMap = new HashMap<>();
+
+        for (String fieldName : schemaFieldMap.keySet()) {
+            Property prop = docModel.getPropertyObject(schemaName, fieldName);
+            if (prop.isDirty()) {
+                fieldMap.put(prop.getName(), prop.getValue());
+            }
+        }
+
+        Document bson = MongoDBSerializationHelper.fieldMapToBson(
+                Collections.singletonMap(getIdField(), docModel.getId()));
+
+        Document props = new Document();
+        props.append(MONGODB_SET, MongoDBSerializationHelper.fieldMapToBson(fieldMap));
```

Comment:
I think this creates the document if it doesn't exist yet, which is not what we want. You should add a unit test for that, and probably change the code to add to the update call a query that basically checks `{idField: { "$exists": true }}` and then verify that the update updated exactly one doc.